### PR TITLE
add `fly deploy --process-groups`

### PR DESCRIPTION
**Pull request by wjordan**
### Change Summary

**What and Why:**
Filters deploy to machines in the specified process groups. Sometimes it's helpful to deploy a targeted release only to a specific subset of process groups in an app (for example, updating/restarting lightweight, stateless process-group services while leaving heavier, stateful process-group services untouched).

**How:**
Adds a `--process-groups` flag to `fly deploy` that sets a `processGroups` field on machineDeployment, which filters the set of machines for deployment as well as the set of process-group names the deploy acts upon (using `md.ProcessNames()` to replace `md.appConfig.ProcessNames()` in a couple spots).

---

### Documentation

- [ ] Fresh Produce
- [ ] In superfly/docs, or asked for help from docs team
- [ ] n/a

### CODE REVIEW

File: internal/command/deploy/machines.go [hunk 1]
```diff
@@ -179,7 +175,14 @@ func NewMachineDeployment(ctx context.Context, args MachineDeploymentArgs) (Mach
 		onlyRegions:            args.OnlyRegions,
 		immediateMaxConcurrent: immedateMaxConcurrent,
 		volumeInitialSize:      volumeInitialSize,
+		processGroups:          args.ProcessGroups,
+	}
+	// TODO: Blend extraInfo into ValidationError and remove this hack
+	if err, extraInfo := md.Validate(ctx); err != nil {
+		fmt.Fprintf(iostreams.FromContext(ctx).ErrOut, extraInfo)
+		return nil, err
```
Comment (dangra) [inline internal/command/deploy/machines.go:183]: why was this block moved?

File: internal/command/deploy/machines.go [hunk 2]
```diff
@@ -103,12 +105,6 @@ func NewMachineDeployment(ctx context.Context, args MachineDeploymentArgs) (Mach
...
 
-	// TODO: Blend extraInfo into ValidationError and remove this hack
-	if err, extraInfo := appConfig.Validate(ctx); err != nil {
-		fmt.Fprintf(iostreams.FromContext(ctx).ErrOut, extraInfo)
-		return nil, err
-	}
-
 	if err != nil {
 		return nil, err
 	}
```
Comment (dangra) [inline internal/command/deploy/machines.go:110]: this must be some leftover code, not related to this PR but can be removed.

File: internal/command/deploy/machines.go [hunk 3]
```diff
@@ -553,3 +566,27 @@ func determineAppConfigForMachines(ctx context.Context, envFromFlags []string, p
...
+	names = md.appConfig.ProcessNames()
+	if len(md.processGroups) > 0 {
+		names = lo.Filter(names, func(name string, _ int) bool {
+			return md.processGroups[name] != nil
+		})
+	}
+	return
+}
+
+// Validate validates app config for process groups in the deployment.
```
Comment (dangra) [inline internal/command/deploy/machines.go:580]: This flattening+validation is something that can happen as part of appConfig.Validate() call. 

It is that or bring the appConfig.GroupConfig() functionality in here

Comment (wjordan) [reply internal/command/deploy/machines.go]: Thanks for the feedback, I took another pass at this in 31a8581a04d9209dbcb5536247f6c9837fc24c51 which moves everything into an `appConfig.ValidateGroups()` call.

File: internal/command/deploy/machines.go (hunk 1 above)
Comment (wjordan) [reply internal/command/deploy/machines.go]: The call to `md.Validate()` needed to be moved after the `md` object was defined, but I moved it back in 31a8581a04d9209dbcb5536247f6c9837fc24c51 which changes the call to `appConfig.ValidateGroups(ctx, lo.Keys(args.ProcessGroups))`.

**Review by dangra — APPROVED** [review]
Approving with a nitpick: One of the "features" of `internal/appconfig` package is that it is relatively well covered by unit tests. Adding one for ValidateGroups() would be awesome.

**Review by dangra — APPROVED** [review]
Good stuff!

**Pull request merged**